## Pandas analysis

This exercise consists in analyzing a dataset containg timing information from a series of Time-to-Digital-Converters (TDC) implemented in a couple of FPGAs. Each measurement (i.e. each row of the input file) consists of a flag that specifies the type of message ('HEAD', which in this case is always 1), two addresses of the TDC providing the signal ('FPGA' and 'TDC_CHANNEL'), and the timing information ('ORBIT_CNT', 'BX_COUNTER', and 'TDC_MEAS'). Each TDC count corresponds to 25/30 ns, whereas a unit of BX_COUNTER corresponds to 25 ns, and the ORBIT_CNT is increased every 'x' BX_COUNTER. This allows to store the time in a similar way to hours, minutes and seconds.

In [1]:
# If you didn't download it yet, please get the relevant file now!
# !wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ~/data/

1\. Create a Pandas DataFrame reading N rows of the 'data_000637.txt' dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k.


In [3]:
import numpy as np
import pandas as pd

In [59]:
!grep -c -v "^#" "/home/ponchiom/data/data_000637.txt"
N = 1310721 -1
df = pd.read_csv("/home/ponchiom/data/data_000637.txt", delimiter=',', nrows = N)
df

1310721


HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0           1     0          123  3869200167        2374        26
1           1     0          124  3869200167        2374        27
2           1     0           63  3869200167        2553        28
3           1     0           64  3869200167        2558        19
4           1     0           64  3869200167        2760        25
5           1     0           63  3869200167        2762         4
6           1     0           61  3869200167        2772        14
7           1     0          139  3869200167        2776         0
8           1     0           62  3869200167        2774        21
9           1     0           60  3869200167        2788         7
10          1     1            7  3869200167        2785         4
11          1     0           64  3869200167        2786        19
12          1     1            6  3869200167        2792        18
13          1     0           36  3869200167        2791        23
14          1     0           56  3869200167        2789         3
15          1     1          139  3869200167        2797         0
16          1     1            8  3869200167        2787        14
17          1     0           63  3869200167        2790        10
18          1     1            5  3869200167        2795         4
19          1     0           53  3869200167        2796        26
20          1     1           10  3869200167        2789        14
21          1     0           57  3869200167        2789        10
22          1     0           61  3869200167        2790        23
23          1     0           38  3869200167        2799        15
24          1     0           58  3869200167        2795        19
25          1     0           62  3869200167        2797        14
26          1     0           59  3869200167        2799        14
27          1     0           59  3869200167        3081        21
28          1     0           61  3869200167        3081         1
29          1     0           60  3869200167        3083        20
...       ...   ...          ...         ...         ...       ...
1310690     1     1            1  3869211170        3517         5
1310691     1     0           56  3869211170        3516        21
1310692     1     1            8  3869211170        3516         7
1310693     1     0           58  3869211170        3521        14
1310694     1     1           12  3869211170        3524        15
1310695     1     1            2  3869211170        3527        24
1310696     1     0           63  3869211170        3526         2
1310697     1     1           10  3869211170        3525         3
1310698     1     0           60  3869211171           9         9
1310699     1     0           25  3869211171          11        28
1310700     1     0           28  3869211171           9        26
1310701     1     0          139  3869211171          21         0
1310702     1     0           27  3869211171          18        17
1310703     1     0           25  3869211171          20        13
1310704     1     0           26  3869211171          19        15
1310705     1     0           94  3869211171          43        22
1310706     1     0           28  3869211171          44        29
1310707     1     0           57  3869211171         137        29
1310708     1     0          124  3869211171         199         9
1310709     1     1           38  3869211171         417         8
1310710     1     1           36  3869211171         427        13
1310711     1     1           39  3869211171         430         0
1310712     1     1           35  3869211171         429         4
1310713     1     0           64  3869211171         758        18
1310714     1     0           60  3869211171         762         2
1310715     1     0           62  3869211171         762        14
1310716     1     1            4  3869211171         763        11
1310717     1     0           64  3869211171         764

2\. Find out the number of BX in a ORBIT (the value 'x').

In [60]:
df.describe()
# 1 BX_COUNTER = 30 TDC MEAS
# 1 ORBIT_CNT = 3564 BX_COUNTER
x = df.BX_COUNTER.max() + 1
print(x)

3564


3\. Find out how much the data taking lasted. You can either make an estimate based on the fraction of the measurements (rows) you read, or perform this check precisely by reading out the whole dataset.

In [61]:
M = df.ORBIT_CNT.max()
m = df.ORBIT_CNT.min()
m_bx = df.BX_COUNTER.iloc[0]
M_bx = df.BX_COUNTER.iloc[N-1]
m_tdc = df.TDC_MEAS.iloc[0]
M_tdc = df.TDC_MEAS.iloc[N-1]
print(m_bx)
print(M_bx)
# time in ns
time = (M-m)*x*25 - (m_bx*25 + m_tdc*25./30) + (M_bx*25 + M_tdc*25./30)
print(f"Time in seconds: {time*10**(-9)}")
# time in h
print(f"Time in minutes: {time*10**(-9)/60}")

2374
762
Time in seconds: 0.9804160933333335
Time in minutes: 0.016340268222222225


4\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information).

In [66]:
df["Time_ns"] = df.ORBIT_CNT*x*25 + df.BX_COUNTER*25 + df.TDC_MEAS*25.0/30.0
df

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
0           1     0          123  3869200167        2374        26   
1           1     0          124  3869200167        2374        27   
2           1     0           63  3869200167        2553        28   
3           1     0           64  3869200167        2558        19   
4           1     0           64  3869200167        2760        25   
5           1     0           63  3869200167        2762         4   
6           1     0           61  3869200167        2772        14   
7           1     0          139  3869200167        2776         0   
8           1     0           62  3869200167        2774        21   
9           1     0           60  3869200167        2788         7   
10          1     1            7  3869200167        2785         4   
11          1     0           64  3869200167        2786        19   
12          1     1            6  3869200167        2792        18   
13          1     0           36  3869200167        2791        23   
14          1     0           56  3869200167        2789         3   
15          1     1          139  3869200167        2797         0   
16          1     1            8  3869200167        2787        14   
17          1     0           63  3869200167        2790        10   
18          1     1            5  3869200167        2795         4   
19          1     0           53  3869200167        2796        26   
20          1     1           10  3869200167        2789        14   
21          1     0           57  3869200167        2789        10   
22          1     0           61  3869200167        2790        23   
23          1     0           38  3869200167        2799        15   
24          1     0           58  3869200167        2795        19   
25          1     0           62  3869200167        2797        14   
26          1     0           59  3869200167        2799        14   
27          1     0           59  3869200167        3081        21   
28          1     0           61  3869200167        3081         1   
29          1     0           60  3869200167        3083        20   
...       ...   ...          ...         ...         ...       ...   
1310690     1     1            1  3869211170        3517         5   
1310691     1     0           56  3869211170        3516        21   
1310692     1     1            8  3869211170        3516         7   
1310693     1     0           58  3869211170        3521        14   
1310694     1     1           12  3869211170        3524        15   
1310695     1     1            2  3869211170        3527        24   
1310696     1     0           63  3869211170        3526         2   
1310697     1     1           10  3869211170        3525         3   
1310698     1     0           60  3869211171           9         9   
1310699     1     0           25  3869211171          11        28   
1310700     1     0           28  3869211171           9        26   
1310701     1     0          139  3869211171          21         0   
1310702     1     0           27  3869211171          18        17   
1310703     1     0           25  3869211171          20        13   
1310704     1     0           26  3869211171          19        15   
1310705     1     0           94  3869211171          43        22   
1310706     1     0           28  3869211171          44        29   
1310707     1     0           57  3869211171         137        29   
1310708     1     0          124  3869211171         199         9   
1310709     1     1           38  3869211171         417         8   
1310710     1     1           36  3869211171         427        13   
1310711     1     1           39  3869211171         430         0   
1310712     1     1           35  3869211171         429         4   
1310713     1     0           64  3869211171         758        18   
1310714     1     0           60  3869211171         762         2   
1310715     1     0

5\. Replace the values (all 1) of the HEAD column randomly with 0 or 1.

In [67]:
import numpy.random as npr

In [79]:
df["HEAD"] = pd.Series(npr.randint(0,2,N),index=df.index)
df

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
0           1     0          123  3869200167        2374        26   
1           1     0          124  3869200167        2374        27   
2           1     0           63  3869200167        2553        28   
3           0     0           64  3869200167        2558        19   
4           0     0           64  3869200167        2760        25   
5           1     0           63  3869200167        2762         4   
6           1     0           61  3869200167        2772        14   
7           0     0          139  3869200167        2776         0   
8           1     0           62  3869200167        2774        21   
9           0     0           60  3869200167        2788         7   
10          1     1            7  3869200167        2785         4   
11          1     0           64  3869200167        2786        19   
12          0     1            6  3869200167        2792        18   
13          1     0           36  3869200167        2791        23   
14          0     0           56  3869200167        2789         3   
15          1     1          139  3869200167        2797         0   
16          0     1            8  3869200167        2787        14   
17          0     0           63  3869200167        2790        10   
18          1     1            5  3869200167        2795         4   
19          1     0           53  3869200167        2796        26   
20          0     1           10  3869200167        2789        14   
21          1     0           57  3869200167        2789        10   
22          1     0           61  3869200167        2790        23   
23          1     0           38  3869200167        2799        15   
24          0     0           58  3869200167        2795        19   
25          1     0           62  3869200167        2797        14   
26          1     0           59  3869200167        2799        14   
27          1     0           59  3869200167        3081        21   
28          1     0           61  3869200167        3081         1   
29          1     0           60  3869200167        3083        20   
...       ...   ...          ...         ...         ...       ...   
1310690     0     1            1  3869211170        3517         5   
1310691     0     0           56  3869211170        3516        21   
1310692     0     1            8  3869211170        3516         7   
1310693     0     0           58  3869211170        3521        14   
1310694     1     1           12  3869211170        3524        15   
1310695     0     1            2  3869211170        3527        24   
1310696     1     0           63  3869211170        3526         2   
1310697     1     1           10  3869211170        3525         3   
1310698     1     0           60  3869211171           9         9   
1310699     1     0           25  3869211171          11        28   
1310700     0     0           28  3869211171           9        26   
1310701     0     0          139  3869211171          21         0   
1310702     0     0           27  3869211171          18        17   
1310703     0     0           25  3869211171          20        13   
1310704     0     0           26  3869211171          19        15   
1310705     1     0           94  3869211171          43        22   
1310706     0     0           28  3869211171          44        29   
1310707     1     0           57  3869211171         137        29   
1310708     1     0          124  3869211171         199         9   
1310709     1     1           38  3869211171         417         8   
1310710     1     1           36  3869211171         427        13   
1310711     0     1           39  3869211171         430         0   
1310712     0     1           35  3869211171         429         4   
1310713     0     0           64  3869211171         758        18   
1310714     1     0           60  3869211171         762         2   
1310715     0     0

6\. Create a new DataFrame that contains only the rows with HEAD=1.

In [82]:
new_df = df[df["HEAD"]>0]
new_df

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
0           1     0          123  3869200167        2374        26   
1           1     0          124  3869200167        2374        27   
2           1     0           63  3869200167        2553        28   
5           1     0           63  3869200167        2762         4   
6           1     0           61  3869200167        2772        14   
8           1     0           62  3869200167        2774        21   
10          1     1            7  3869200167        2785         4   
11          1     0           64  3869200167        2786        19   
13          1     0           36  3869200167        2791        23   
15          1     1          139  3869200167        2797         0   
18          1     1            5  3869200167        2795         4   
19          1     0           53  3869200167        2796        26   
21          1     0           57  3869200167        2789        10   
22          1     0           61  3869200167        2790        23   
23          1     0           38  3869200167        2799        15   
25          1     0           62  3869200167        2797        14   
26          1     0           59  3869200167        2799        14   
27          1     0           59  3869200167        3081        21   
28          1     0           61  3869200167        3081         1   
29          1     0           60  3869200167        3083        20   
31          1     0           62  3869200167        3079         4   
35          1     0           51  3869200167        3146        20   
36          1     0           59  3869200167        3176         8   
37          1     0           53  3869200167        3181         8   
41          1     0           56  3869200167        3177        20   
42          1     0           57  3869200167        3187        17   
47          1     1            2  3869200168          17        26   
48          1     0          139  3869200168          23         0   
49          1     0           61  3869200168          18        17   
58          1     1          127  3869200168         273         6   
...       ...   ...          ...         ...         ...       ...   
1310658     1     0           59  3869211170        3024        22   
1310659     1     1           17  3869211170        3030         8   
1310660     1     0          139  3869211170        3027         0   
1310661     1     0           63  3869211170        3023        28   
1310662     1     0           42  3869211170        3025        29   
1310664     1     0           49  3869211170        3029         8   
1310665     1     0           60  3869211170        3026        24   
1310667     1     0           41  3869211170        3033        27   
1310668     1     0           47  3869211170        3033        19   
1310669     1     0           50  3869211170        3029        27   
1310671     1     0           37  3869211170        3034        11   
1310672     1     0           48  3869211170        3035        13   
1310676     1     0           62  3869211170        3032         3   
1310683     1     0           43  3869211170        3220        19   
1310685     1     0          139  3869211170        3230         0   
1310688     1     1          109  3869211170        3421        12   
1310689     1     0           28  3869211170        3507        18   
1310694     1     1           12  3869211170        3524        15   
1310696     1     0           63  3869211170        3526         2   
1310697     1     1           10  3869211170        3525         3   
1310698     1     0           60  3869211171           9         9   
1310699     1     0           25  3869211171          11        28   
1310705     1     0           94  3869211171          43        22   
1310707     1     0           57  3869211171         137        29   
1310708     1     0          124  3869211171         199         9   
1310709     1     1

7\. Make two occupancy plots (one for each FPGA), i.e. plot the number of counts per TDC channel

8\. Use the groupby method to find out the noisy channels, i.e. the TDC channels with most counts (say the top 3)

9\. Count the number of unique orbits. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139